### ISSUE TITLE
tensorflow distribute training error: tf.Tensor 'report_uninitialized_variables/boolean_mask/Gather:0' shape=(?,)

### ISSUE THREAD

**Issue by pickou** [tensorflow/models#2483]
### System information
- **What is the top-level directory of the model you are using**: research/inception
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
YES
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.3.0
- **CUDA/cuDNN version**:8.0
- **GPU model and memory**:GeForce GTX TITAN 12GB

### Describe the problem
1. I encountered a problem running the distributed [cifar10_sync_dist_train.py](https://github.com/pickou/cifar10-distribute-latest/blob/master/cifar10_sync_dist_train.py) on tensorflow version 1.3,error occured after the training was done and was about to save the graph.
2. When I use the command as follows to run the distributed [cifar10_sync_dist_train.py](https://github.com/pickou/cifar10-distribute-latest/blob/master/cifar10_sync_dist_train.py) locally, the training is done, but I got an error. 
3. The original [code](https://github.com/caicloud/tensorflow-demo/blob/master/cifar10/cifar10_sync_dist_train.py) can run properly on tf version 0.12.
4. This is the  [code](https://github.com/pickou/cifar10-distribute-latest) I used in tf version 1.3. I only modified the summary related code to fit for the latest version.
### Command
```
python cifar10_sync_dist_train.py --ps_hosts localhost:2220 --job_name ps --worker_hosts localhost:2221,localhost:2222 --task_id 0

python cifar10_sync_dist_train.py --ps_hosts localhost:2220 --job_name worker --worker_hosts localhost:2221,localhost:2222 --task_id 0

python cifar10_sync_dist_train.py --ps_hosts localhost:2220 --job_name worker --worker_hosts localhost:2221,localhost:2222 --task_id 1
```
### ERROR
```bash
ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Tensor'>):
<tf.Tensor 'report_uninitialized_variables/boolean_mask/Gather:0' shape=(?,) dtype=string>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "cifar10_sync_dist_train.py", line 179, in <module>\n    tf.app.run()', 'File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/platform/app.py", line 48, in run\n    _sys.exit(main(_sys.argv[:1] + flags_passthrough))', 'File "cifar10_sync_dist_train.py", line 176, in main\n    train()', 'File "cifar10_sync_dist_train.py", line 113, in train\n    apply_gradients_op = opt.apply_gradients(grads, global_step=global_step)', 'File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/training/sync_replicas_optimizer.py", line 257, in apply_gradients\n    variables.global_variables())', 'File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/util/tf_should_use.py", line 170, in wrapped\n    return _add_should_use_warning(fn(*args, **kwargs))', 'File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/util/tf_should_use.py", line 139, in _add_should_use_warning\n    wrapped = TFShouldUseWarningWrapper(x)', 'File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/util/tf_should_use.py", line 96, in __init__\n    stack = [s.strip() for s in traceback.format_stack()]']
```

### Source code / logs
Here is the code I use.
https://github.com/caicloud/tensorflow-demo/blob/master/cifar10/cifar10_sync_dist_train.py

**Comment by cy89** [comment]
Do the other bugs that include "object was never used" shed any light?
https://github.com/tensorflow/tensorflow/issues/11315
https://github.com/openai/pixel-cnn/issues/17

It seems unsurprising that you're getting the error on more recent versions--it sounds like this is an error checking feature, where you might have built something but never invoked it in your custom code.

**Comment by pickou** [comment]
I've tried this before. Since I have no learning rate declared and keep_prob as constant, I'll check the others .

**Comment by pickou** [comment]
I have checked the code carefully and I couldn't find anything built but never used. I've checked the global_variables_initializer and local_variables_initializer, not work. And I also try use mark_used() to init_op, but turned out it had no such method. So I followed the error 

```'File "cifar10_sync_dist_train.py", line 113, in train\n    apply_gradients_op = opt.apply_gradients(grads, global_step=global_step)', 'File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/training/sync_replicas_optimizer.py", line 257, in apply_gradients\n    variables.global_variables())'```, 
and I found this line
 ```self.ready_for_local_init_op = variables.report_uninitialized_variables(variables.global_variables())```
So, I believe something went wrong when initializing the variables.

**Comment by reedwm** [comment]
Closing, as this is a duplicate of tensorflow/tensorflow#11823. Pinging @ali01 on that issue.
